sam/bamWrite.go

```diff
@@ -0,0 +1,349 @@
+package sam
+
+import (
+	"bytes"
+	"encoding/hex"
+	"github.com/vertgenlab/gonomics/bgzf"
+	"github.com/vertgenlab/gonomics/cigar"
+	"github.com/vertgenlab/gonomics/exception"
+	"io"
+	"log"
+	"strconv"
+	"strings"
+)
+
+// BamWriter wraps a bgzf.Writer and provides functions to write
+// a Sam struct to a binary Bam file, including WriteToBamFileHandle.
+type BamWriter struct {
+	bgzf.Writer
+	buf       bytes.Buffer // for storage between records
+	recordBuf bytes.Buffer // for use within 1 record
+	refMap    map[string]int
+	u32       [4]byte // bytes for writing little endian
+}
+
+const (
+	nul byte = '\u0000'
+)
+
+// NewBamWriter creates a new bam writer to the input io.Writer.
+// The magic bam bytes and header are immediately written to the BamWriter.
+func NewBamWriter(w io.Writer, h Header) *BamWriter {
+	var bw BamWriter
+	bw.Writer = bgzf.NewWriter(w)
+	bw.buf.WriteString(magicBam)
```

Comment:
Where does this magicBam come from? And so I understand this line, will the output from the WriteString() be placed in  the bw variables struct, in the buf position? I'm also not sure where this WriteString is coming from?